First draft of action definition for scanning actions.

airballking opened the pull request:
@ichumuh @amaldo Could you please have a look at this action definition, and say whether it seems reasonable to you?

ichumuh commented on action/Scanning.action:
```diff
@@ -0,0 +1,33 @@
+int8 SEPARATOR=3  # detect and localize separators
+int8 OBJECT_RECOGNITION=4  # detect, recognize, and localize front object in facing
+int8 OBJECT_COUNTING=5  # count all objects within one facing
+
```
This definition does not allow us to do a full store inventory with one command. And if we want split it up and do it in the most efficient way, we'd need some kind of high level outside of the action server. Is this intended?

ichumuh commented on action/Scanning.action:
```diff
@@ -0,0 +1,33 @@
+# Result definition of a single scanning action.
+
+# Pre-defined constants to code possible runtime failures.
+int8 SUCCESS=-1  # to report successful execution
```
is SUCCESS=-1 common practice? How about SUCCESS=0 and error codes are 1000<=error_code<=9999?

ichumuh commented on action/Scanning.action:
```diff
@@ -0,0 +1,33 @@
+
+---
+# Feedback definition of a single scanning action.
+refills_msgs/Location current_loc  # location currently being scanned
```
the problem with using location.msg in feedback is that it only contains rack OR shelf OR floor etc. but we probably want to know current rack AND shelf AND floor etc right?

airballking replied:
I do not think so. I imagine this to be a KnowRob ID. If the ID of a shelf floor is returned, then the recipient can query KnowRob to figure out to which shelf or rack this shelf floor belongs, right?

ichumuh commented on msg/Location.msg:
```diff
@@ -0,0 +1,23 @@
+# this location.
+# Example: For queries like 'Scan rack-42' set 'use_loc_id=True'.
+# Example: For queries like 'Scan rack with product XY' set 'use_loc_id=False'.
+bool use_loc_id
```
use_loc_id -> use_product_id because I think that should be the default

ichumuh commented on msg/Location.msg:
```diff
@@ -0,0 +1,23 @@
+string loc_id
+
+# Unique of product classes, as defined in semantic store maps.
+string product_id
```
- is type used in combination with product_id? It might be best to not force both functionalities in to one msg/action?
- how do I know where products are? We might have some knowledge from previous scans but to be certain the robot has to scan everything or at least a full shelf?

airballking commented:
@ichumuh Thank you for your review! I also talked to @amaldo about the suggested interface, yesterday. He mirrored some of your concerns. In general, he was advocating for a simpler interface. 

Specifically, he suggested (1) less choices for the type of scanning action, with a default of 'complete_scan', (2) the smallest represented location should be a 'shelf', neither a 'shelf floor' nor a 'facing', and (3) he does not like the 'product_id' option --that functionality can be encapsulated in the GUI program.

I will revise the definitions and update this PR. Read you soon!

airballking commented:
@amaldo and @ichumuh here is a revised version of ```Scanning.action```. Could you please have a look? Basically, I removed the extra message definition ```Location.msg``` and removed a lot of the type switches.

ichumuh commented on action/Scanning.action:
```diff
@@ -2,24 +2,21 @@
-refills_msgs/Location[] locations
+# List of locations that shall be scanned, encoded by their unique IDs,
+# as defined in the semantic store maps.
+string[] loc_id
```
with location you mean shelf or can it also refer to racks?

airballking replied:
Correct. I extended the comment to be more specific.

airballking commented:
@ichumuh and @amaldo Can we now merge this as a first simplistic interface definition?

airballking commented:
@ichumuh and @amaldo: Ping. ;)

ichumuh commented:
Yes I think its fine for a first draft.
